source/azure_iot_hub_client_properties.c

```diff
@@ -0,0 +1,222 @@
+{
+    AzureIoTResult_t xResult;
+    az_result xCoreResult;
+    az_iot_hub_client_properties_response_type xCoreResponseType;
+
+    if( ( pxAzureIoTHubClient == NULL ) || ( pxJSONReader == NULL ) ||
+        ( ( xResponseType != eAzureIoTHubPropertiesGetMessage ) && ( xResponseType != eAzureIoTHubPropertiesWriteablePropertyMessage ) ) ||
+        ( pulVersion == NULL ) )
+    {
+        AZLogError( ( "AzureIoTHubClientProperties_GetPropertiesVersion failed: invalid argument" ) );
+        return eAzureIoTErrorInvalidArgument;
+    }
+
+    xCoreResponseType = xResponseType == eAzureIoTHubPropertiesGetMessage ?
+                        AZ_IOT_HUB_CLIENT_PROPERTIES_RESPONSE_TYPE_GET : AZ_IOT_HUB_CLIENT_PROPERTIES_RESPONSE_TYPE_DESIRED_PROPERTIES;
+
+    if( az_result_failed(
+            xCoreResult = az_iot_hub_client_properties_get_properties_version( &pxAzureIoTHubClient->_internal.xAzureIoTHubClientCore,
+                                                                               &pxJSONReader->_internal.xCoreReader, xCoreResponseType, ( int32_t * ) pulVersion ) ) )
+    {
+        AZLogError( ( "Could not get property version: core error=0x%08x", xCoreResult ) );
+        xResult = eAzureIoTErrorFailed;
+    }
+    else
+    {
+        xResult = eAzureIoTSuccess;
+    }
+
+    return xResult;
+}
+
+AzureIoTResult_t AzureIoTHubClientProperties_GetNextComponentProperty( AzureIoTHubClient_t * pxAzureIoTHubClient,
+                                                                       AzureIoTJSONReader_t * pxJSONReader,
+                                                                       AzureIoTHubMessageType_t xResponseType,
+                                                                       AzureIoTHubClientPropertyType_t xPropertyType,
+                                                                       const uint8_t ** ppucComponentName,
+                                                                       uint32_t * pulComponentNameLength )
+{
+    AzureIoTResult_t xResult;
+    az_result xCoreResult;
+    az_span xComponentSpan;
+    az_iot_hub_client_properties_response_type xCoreResponseType;
+
+    if( ( pxAzureIoTHubClient == NULL ) || ( pxJSONReader == NULL ) ||
+        ( ( xResponseType != eAzureIoTHubPropertiesGetMessage ) && ( xResponseType != eAzureIoTHubPropertiesWriteablePropertyMessage ) ) ||
+        ( ppucComponentName == NULL ) || ( pulComponentNameLength == NULL ) )
+    {
+        AZLogError( ( "AzureIoTHubClientProperties_GetNextComponentProperty failed: invalid argument" ) );
+        return eAzureIoTErrorInvalidArgument;
+    }
+
+    xComponentSpan = az_span_create( ( uint8_t * ) *ppucComponentName, ( int32_t ) *pulComponentNameLength );
+    xCoreResponseType = xResponseType == eAzureIoTHubPropertiesGetMessage ?
+                        AZ_IOT_HUB_CLIENT_PROPERTIES_RESPONSE_TYPE_GET : AZ_IOT_HUB_CLIENT_PROPERTIES_RESPONSE_TYPE_DESIRED_PROPERTIES;
+
+    if( az_result_failed(
```

Comment:
The underlying embedded API is going to return `AZ_ERROR_IOT_END_OF_PROPERTIES` when there's no more properties.  However this is going to end up masking this legit status into an `eAzureIoTErrorFailed`.  I think we need a new status code for `AzureIoTResult_t` and to special case transforming that specific error back to caller.

Reply:
Really good catch thank you. Updated

Reply:
Added to the ut as well to make sure we hit that:
https://github.com/Azure/azure-iot-middleware-freertos/pull/135/commits/0d00447bb19ec12fd1dafa976a81a948c553312b